kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -84,16 +84,99 @@ public class TransactionImpl implements Transaction {
   private final List<Column> clusteringColumns;
   private final boolean shouldUpdateProtocol;
   private final Clock clock;
-  private final Map<String, DomainMetadata> domainMetadatasAdded = new HashMap<>();
-  private final Set<String> domainMetadatasRemoved = new HashSet<>();
-  private Optional<List<DomainMetadata>> domainMetadatas = Optional.empty();
+  private final DomainMetadataState domainMetadataState = new DomainMetadataState();
   private Metadata metadata;
   private boolean shouldUpdateMetadata;
   private int maxRetries;
   private Optional<CRCInfo> currentCrcInfo;
 
   private boolean closed; // To avoid trying to commit the same transaction again.
 
+  /** Encapsulates the state of domain metadata within a transaction. */
+  private class DomainMetadataState {
```

Comment:
Looks great! Please move this to the end of the file (and, yes, we should develop a dev guide where order and best practices like this are stated)